agent/consul/acl.go

```diff
@@ -660,39 +569,22 @@ func (r *ACLResolver) resolvePoliciesForIdentity(identity structs.ACLIdentity) (
 		return r.filterPoliciesByScope(policies), nil
 	}
 
+	hasMissing := len(missing) > 0
+
 	fetchIDs := missing
 	for _, policy := range expired {
 		fetchIDs = append(fetchIDs, policy.ID)
 	}
 
 	// Background a RPC request and wait on it if we must
-	var waitChan chan *remoteACLPolicyResult
-	waitForResult := len(missing) > 0 || r.config.ACLDownPolicy != "async-cache"
-	if waitForResult {
-		// buffered because there are going to be multiple go routines that send data to this chan
-		waitChan = make(chan *remoteACLPolicyResult, len(fetchIDs))
-	}
-
-	var newAsyncFetchIds []string
-	r.asyncPolicyResultsMutex.Lock()
-	for _, policyID := range fetchIDs {
-		clients, ok := r.asyncPolicyResults[policyID]
-		if !ok || clients == nil {
-			clients = make([]chan *remoteACLPolicyResult, 0)
-		}
-		if waitForResult {
-			r.asyncPolicyResults[policyID] = append(clients, waitChan)
-		}
-
-		if !ok {
-			newAsyncFetchIds = append(newAsyncFetchIds, policyID)
-		}
-	}
-	r.asyncPolicyResultsMutex.Unlock()
+	var waitChan <-chan singleflight.Result
 
-	if len(newAsyncFetchIds) > 0 {
-		// only start the RPC if one isn't in flight
-		go r.resolvePoliciesAsyncForIdentity(identity, newAsyncFetchIds, expCacheMap)
+	waitForResult := hasMissing || r.config.ACLDownPolicy != "async-cache"
+	if len(fetchIDs) > 0 {
```

Comment:
Is this length check still needed? 

Before the length check was there to handle single flighting. Couldn't we just unconditionally do the DoChan call now?

Reply:
Yeah that can go now. I'd missed that I'd deleted the code that could make that len check be zero.